### PR TITLE
refactor(app): add protocol run details setup tab and robot calibration

### LINKED ISSUE #8748: Setup tab - Step 1 (Robot Calibration)
# Overview
The protocol run details page has a Setup tab with 3 accordion sections. The first section is related to Robot Calibration. 

# Design
Here are the various states of the Step 1 accordion: 
- Step 1 accordion collapsed + calibration needed: https://www.figma.com/file/bVtCogdwiXj3ndpk4AWXCx/%5Bworking%5D-Milestone-1%3A-Read-and-Execute?node-id=1297%3A101653
- Step 1 accordion collapsed + calibration ready: https://www.figma.com/file/bVtCogdwiXj3ndpk4AWXCx/%5Bworking%5D-Milestone-1%3A-Read-and-Execute?node-id=1297%3A101653
- Step 1 accordion expanded + calibration ready: https://www.figma.com/file/bVtCogdwiXj3ndpk4AWXCx/%5Bworking%5D-Milestone-1%3A-Read-and-Execute?node-id=1297%3A101653
- Step 1 accordion expanded + calibration needed: https://www.figma.com/file/bVtCogdwiXj3ndpk4AWXCx/%5Bworking%5D-Milestone-1%3A-Read-and-Execute?node-id=1297%3A101653

# Acceptance Criteria

- [ ] If the robot calibration is not complete, the "Start Run" button should be inactive
- [ ] If the robot calibration is complete, the "Start Run" button should be active
- [ ] Clicking on any of the calibration links should bring the user to the correct calibration workflow
- [ ] Clicking on "Proceed to Module Setup" should collapse the Step 1 accordion and expand the Step 2 accordion

# Selector for E2E

- [ ] Start Run button

### PR DESCRIPTION
**Pull request by brenthagen**
# Overview

Screen Shot 2022-04-01 at 2 10 53 PM

this adds the protocol run details setup tab and the robot calibration section. The old legacy modals for pipette and tip length calibration are reused. there are a few other updates:

 - splits out the `LegacyApp` from the root app component
 - memoizes the selector called by `useRobot` to reduce unnecessary rerenders throughout the unified app
 - reverts the default background color of the legacy `Overlay` component to `rgba(0, 0, 0, 0.9)`, because some old modals are difficult to read in both old and new apps: 

current:
Screen Shot 2022-03-30 at 11 44 23 AM

updated (reverted):

Screen Shot 2022-04-01 at 2 26 32 PM

I adjusted the `Slideout` component to keep the existing overlay background color.

closes #8748

# Changelog

 - Adds protocol run details setup tab and robot calibration
 - Reverts default overlay background color

# Review requests

manual testing of various pipette and tip length calibration states, if possible. check and confirm modals/slideouts/etc are ok

# Risk assessment

low

### CODE REVIEW

File: app/src/atoms/Slideout/index.tsx [hunk 1]
```diff
@@ -75,7 +75,9 @@ export const Slideout = (props: Props): JSX.Element | null => {
   } = props
   return (
     <>
-      {isExpanded ? <Overlay /> : null}
+      {isExpanded ? (
+        <Overlay backgroundColor={`${COLORS.darkBlack}59`} />
```
Comment (b-cooper) [inline app/src/atoms/Slideout/index.tsx]: Can we pull this out into a local variable with an explanatory comment?  Ideally we'd have this color with an alpha value saved as another exported COLOR const.

File: app/src/organisms/Devices/ProtocolRun/SetupDeckCalibration.tsx [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+export function SetupDeckCalibration({
+  robotName,
+}: SetupDeckCalibrationProps): JSX.Element | null {
+  const { t } = useTranslation('protocol_setup')
+
+  const deckCalData = useDeckCalibrationData(robotName)
+
+  // this component should never be rendered if there is no deckCalData
```
Comment (b-cooper) [inline app/src/organisms/Devices/ProtocolRun/SetupDeckCalibration.tsx]: This comment is actually not true anymore, based on the current structure of the app. This means that we will probably want to expand this calibration setup step slightly to allow a user to be linked to complete deck calibration from the robot setting page.  For now, if deck cal is missing here, could you add in a nav link to the robot settings area where deck cal can be accessed. The copy should say something similar to what we say for the two other required calibrations "Please perform Deck Calibration before ..." @emilywools maybe you can help with this copy?

File: app/src/organisms/Devices/ProtocolRun/SetupStep.tsx [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  expanded: boolean
+  title: React.ReactNode
+  description: string
+  label: string
+  toggleExpanded: () => void
+  children: React.ReactNode
+  calibrationStatusComplete: boolean | null
+}
+
+const EXPANDED_STYLE = css`
```
Comment (b-cooper) [inline app/src/organisms/Devices/ProtocolRun/SetupStep.tsx:34]: Small detail for a follow up: Something about this collapse transition broke here.  When I collapse the robot calibration tab, I should see the contents of the becoming obscured by the "closing" container, and then it's visibility should be set to hidden only after it has already been fully obscured (this part is necessary to test which item is open in unit tests).  

When I tested this out, the step's contents hid immediately, and then the step collapsed as an empty white div.

File: app/src/organisms/Devices/ProtocolRun/SetupTipLengthCalibrationButton.tsx [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+import type { LabwareDefinition2 } from '@opentrons/shared-data'
+import type { RequestState } from '../../../redux/robot-api/types'
+import type {
+  SessionCommandString,
+  PipetteOffsetCalibrationSession,
+  TipLengthCalibrationSession,
+} from '../../../redux/sessions/types'
+import type { State } from '../../../redux/types'
+
+const TIP_LENGTH_CALIBRATION = 'tip length calibration'
```
Comment (b-cooper) [inline app/src/organisms/Devices/ProtocolRun/SetupTipLengthCalibrationButton.tsx]: i18n

File: app/src/organisms/Devices/ProtocolRun/SetupTipLengthCalibrationButton.tsx [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+  const trackedRequestId = React.useRef<string | null>(null)
+  const jogRequestId = React.useRef<string | null>(null)
+  const dispatch = useDispatch()
+  const { t } = useTranslation('protocol_setup')
+
+  const sessionType = isExtendedPipOffset
+    ? Sessions.SESSION_TYPE_PIPETTE_OFFSET_CALIBRATION
+    : Sessions.SESSION_TYPE_TIP_LENGTH_CALIBRATION
+
+  const uriFromDef = (definition: LabwareDefinition2): string =>
```
Comment (b-cooper) [inline app/src/organisms/Devices/ProtocolRun/SetupTipLengthCalibrationButton.tsx]: I believe this function already exists in shared-data as `getLabwareDefURI`

File: app/src/redux/discovery/selectors.ts [hunk 6]
```diff
@@ -37,6 +38,9 @@ type GetAllRobots = (state: State) => DiscoveredRobot[]
 type GetViewableRobots = (state: State) => ViewableRobot[]
 type GetConnectedRobot = (state: State) => Robot | null
 
+// from https://github.com/reduxjs/reselect#customize-equalitycheck-for-defaultmemoize
+const createDeepEqualSelector = createSelectorCreator(defaultMemoize, isEqual)
```
Comment (b-cooper) [inline app/src/redux/discovery/selectors.ts:42]: 🙌

File: shared-data/protocol/fixtures/6/simpleV6.json [hunk 7]
```diff
@@ -1254,6 +1254,9 @@
       "params": {
         "pipetteId": "pipetteId",
         "mount": "left"
+      },
```
Comment (b-cooper) [inline shared-data/protocol/fixtures/6/simpleV6.json]: is this addition to the fixture here necessary?   we're going to need to do an audit of these fixtures to tease apart "protocol analysis output" from "protocol creation output". At the moment I believe the other fixtures in this dir are meant to be "protocol creation output" only. This addition breaks that assumption

**Review by b-cooper — APPROVED** [review]
Code looks good 📦 .  Testing things out and they look great too!

I dropped a few comments in. Overall, I'd like to see these new components living at the top level of the `organisms` directory where possible. Having sub-components nested within `organisms/Devices/**` defeats the purpose of considering every component in `app/src/organisms/` it's own fully qualified "portable" chunk.  This doesn't have to be done in this PR, but we should probably make a pass next week to rearrange some of these components and start to mark the legacy components as deprecated for ease of removal.

File: app/src/organisms/Devices/ProtocolRun/SetupStep.tsx (hunk 3 above)
Comment (brenthagen) [reply app/src/organisms/Devices/ProtocolRun/SetupStep.tsx:34]: same appears to be true in 5.0. i think adding `visibility 600ms ease` to the collapsed style transition will fix.

File: shared-data/protocol/fixtures/6/simpleV6.json (hunk 7 above)
Comment (brenthagen) [reply shared-data/protocol/fixtures/6/simpleV6.json]: I probably need to grab a different fixture for analysis output - this was paired with the change above in useRunPipetteInfoByMount from `command.params` to `command.result`, which was I think based on protocol analysis output from the robot.

Comment (b-cooper) [reply shared-data/protocol/fixtures/6/simpleV6.json]: I've just added a fixture for analysis output in my PR, we can converge after both merge.

Comment (brenthagen) [reply shared-data/protocol/fixtures/6/simpleV6.json]: 👀 i saw that fixture and am eagerly awaiting

File: app/src/organisms/Devices/ProtocolRun/SetupDeckCalibration.tsx (hunk 2 above)
Comment (brenthagen) [reply app/src/organisms/Devices/ProtocolRun/SetupDeckCalibration.tsx]: @emilywools created a set of "no deck cal" designs that I am following: https://www.figma.com/file/bVtCogdwiXj3ndpk4AWXCx/Milestone-1%3A-Read-and-Execute?node-id=8790%3A260369

**Pull request merged**